Assets/Mirror/Runtime/ClientObjectManager.cs

```diff
@@ -0,0 +1,685 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+using UnityEngine.SceneManagement;
+using Guid = System.Guid;
+using Object = UnityEngine.Object;
+
+namespace Mirror
+{
+    /// <summary>
+    /// This is a client object manager class used by the networking system. It manages spawning and unspawning of objects for the network client.
+    /// </summary>
+    [RequireComponent(typeof(NetworkClient))]
+    [DisallowMultipleComponent]
+    public class ClientObjectManager : MonoBehaviour
+    {
+        static readonly ILogger logger = LogFactory.GetLogger(typeof(NetworkClient));
+
+        public NetworkClient client;
+
+        // spawn handlers
+        readonly Dictionary<Guid, SpawnHandlerDelegate> spawnHandlers = new Dictionary<Guid, SpawnHandlerDelegate>();
+        readonly Dictionary<Guid, UnSpawnDelegate> unspawnHandlers = new Dictionary<Guid, UnSpawnDelegate>();
+
+        /// <summary>
+        /// List of prefabs that will be registered with the spawning system.
+        /// <para>For each of these prefabs, ClientManager.RegisterPrefab() will be automatically invoke.</para>
+        /// </summary>
+        public List<GameObject> spawnPrefabs = new List<GameObject>();
+
+        bool isSpawnFinished;
+
+        /// <summary>
+        /// This is a dictionary of the prefabs that are registered on the client with ClientScene.RegisterPrefab().
+        /// <para>The key to the dictionary is the prefab asset Id.</para>
+        /// </summary>
+        private readonly Dictionary<Guid, GameObject> prefabs = new Dictionary<Guid, GameObject>();
+
+        /// <summary>
+        /// This is dictionary of the disabled NetworkIdentity objects in the scene that could be spawned by messages from the server.
+        /// <para>The key to the dictionary is the NetworkIdentity sceneId.</para>
+        /// </summary>
+        public readonly Dictionary<ulong, NetworkIdentity> spawnableObjects = new Dictionary<ulong, NetworkIdentity>();
+
+        public void Start()
+        {
+            if (logger.LogEnabled()) logger.Log("ClientObjectManager started");
+
+            SceneManager.sceneLoaded += OnSceneLoaded;
+
+            client.Connected.AddListener(Connected);
+            client.Disconnected.AddListener(Disconnected);
+        }
+
+        public void OnValidate()
+        {
+            // add NetworkClient if there is none yet. makes upgrading easier.
+            if (GetComponent<NetworkClient>() == null)
```

Comment:
Is it really needed? We're upgrading from a NetworkClient, not a ClientObjectManager. Also, RequireComponent would prevent users to add a ClientObjectManager without a NetworkClient component anyway

Reply:
I was not sure what would 'feel' right for usage with comp requirements. I personally dont like OnValidate all that much since it doesn't cover all cases. I would rather use require comp or be responsible to manually wire it.

Reply:
yeah,  let's get rid of OnValidate,  we can always do:
```
client = client ?? Getcomponent<NetworkClient>()
```
on start

Reply:
fixed per Pauls suggestion